### PR TITLE
Use vscode workspace service

### PR DESCRIPTION
**Pull request by CGNonofr**
This MR is a breaking change:
- `Services.install({ workspace })` doesn't exist anymore. Instead, the default workspace folder can be provided to `getConfigurationServiceOverride` and updated through the vscode api (`vscode.workspace.updateWorkspaceFolders`)
- `StandaloneServices.initialize` shouldn't be used directly but `initialize` from `vscode/services` instead (and the new method retuns a promise that should be awaited before creating an editor and interacting with monaco)
- `initialize` from `vscode/extensions` should be called and awaited before interacting with the vscode api. It can take the default extension as parameter to override it
- `getJsonSchemas` was replaced by `synchronizeJsonSchemas` to simplify usage

These changes allow to remove most of the hacks that were due to the poor lifecycle management we had there, now it's much more close to how vscode works

What it adds:
- it registers a memory filesystem on the `file` scheme
- it exposes a `createModelReference` method that creates a model reference from a file on the memory filesystem (the method take a second argument with the file content to be written to the memory filesystem before creating the model); this method is already called internally a lot (on a ctrl+click with LSP for instance) and vscode is not designed to have model not created via a reference (it can no longer count references or even know there is an existing instance)
- `configuration` and `configurationDefaults` from extension manifests are now used
- a `registerFileSystemOverlay` method is exposed (from `service-override/files`) to be able to plug a remote filesystem. It also exposed a `SimpleTextFileSystemProvider` class for simple usages. Note that there is no a filesystem registered on `file` so we can't use `registerTextModelContentProvider on `file` anymore)
- it implements more of the `vscode.workspace` api

It's another BIG PR, sorry for that

note that monaco 0.37 just got released, and since there is some breaking changes here, we can chose to no release it and wait for the update to monaco 0.37/vscode 1.78 that will follow

### CODE REVIEW

**Comment by kaisalmen** [comment]
Sounds good, but it's a lot of code again.

> note that monaco 0.37 just got released, and since there is some breaking changes here, we can chose to no release it and wait for the update to monaco 0.37/vscode 1.78 that will follow

It makes sense to me and I have more time to review it. 🙈  vscode release is after the 28th this month according to their plans.
Can you prevent semantic release, because then we could merge this in and have the 0.37 afterwards and then got to 1.78.0 once it is available.

**Comment by CGNonofr** [comment]
> Sounds good, but it's a lot of code again.

You're right ! It's the last big refactor until a long time I hope :crossed_fingers: 

> > note that monaco 0.37 just got released, and since there is some breaking changes here, we can chose to no release it and wait for the update to monaco 0.37/vscode 1.78 that will follow
> 
> It makes sense to me and I have more time to review it. see_no_evil vscode release is after the 28th this month according to their plans. Can you prevent semantic release, because then we could merge this in and have the 0.37 afterwards and then advance to the version to 1.78.0 once next vscode is available.

I can just cancel the release pipeline after merging, I already did it

File: src/service-override/files.ts [hunk 1]
```diff
@@ -55,9 +54,82 @@ class File implements IStat {
   }
 }
 const encoder = new TextEncoder()
-class ExtensionFileSystemProviderWithFileReadWriteCapability implements IFileSystemProviderWithFileReadWriteCapability {
+const decoder = new TextDecoder()
+
+abstract class SimpleTextFileSystemProvider implements IFileSystemProviderWithFileReadWriteCapability {
```
Comment (kaisalmen) [inline src/service-override/files.ts:59]: Nice and the application in the demo!

File: src/service-override/files.ts [hunk 2]
```diff
@@ -81,47 +153,182 @@ class ExtensionFileSystemProviderWithFileReadWriteCapability implements IFileSys
...
   constructor (@ILogService logService: ILogService) {
     super(logService)
 
-    this.registerProvider('user', new InMemoryFileSystemProvider())
+    const userMemoryFileSystem = new InMemoryFileSystemProvider()
+
+    this.registerProvider('user', userMemoryFileSystem)
     this.registerProvider('extension', extensionFileSystemProvider)
+    this.registerProvider('cache', new InMemoryFileSystemProvider())
+    this.registerProvider('file', fileSystemProvider)
```
[the remark this answers is not in the document]
Comment (kaisalmen) [reply src/service-override/files.ts:301]: It is written in the description already 🙂

**Review by kaisalmen — APPROVED** [review]
Good job.
I was wondering if it makes sense to introduce unit test using vitest for testing the services instead of having only the demo for verification.

**Comment by CGNonofr** [comment]
> Good job. I was wondering if it makes sense to introduce unit test using vitest for testing the services instead of having only the demo for verification.

Yep! everything that helps preventing regressions is worthwhile. I'm not sure how to do that though...

**Comment by kaisalmen** [comment]
I have some ideas, but it will be likely be May before I come back to it.

**Comment by kaisalmen** [comment]
Something like a monaco-editor/monaco-vscode-api test jumpstart / boilerplate kit that let's you then help to write specific tests shorter.

**Pull request merged**